---
title: "Chapter 4: Single-cell sequencing applications"
output: html_document
date: "2023-08-07"
---

```{r setup, include=FALSE}

knitr::opts_chunk$set(echo = TRUE)

```

**Import data**
```{r}

nice_table <- read.csv("Output files/nice_table.csv")
nice_table2 <- read.csv("Output files/nice_table2.csv")
single_cell <- read.csv("Input files/Single cell seq.csv")
cogvspath_significant <- read.csv("Output files/cogvspath_significant.csv")
cognition_cont_significant <- read.csv("Output files/cognition_cont_significant.csv")

```

**Group sequencing data by cell-type, filter proteins associated with cognition to plot for normalised to all cell-types**
```{r}

single_cell_again <- single_cell %>%
  slice(-(1:4)) %>% #get rid of first 4 rows (not applicable)
  mutate(RowSum = rowSums(.[, -1])) %>%
  filter(RowSum != 0.000) %>%  #get rid of rows that have 0 for all cell types across a gene (doesn't mean unimportant just better to plot)
  mutate(RowMean = RowSum/121) %>% #take mean
  mutate(across(-c(feature, RowMean, RowSum), ~ ifelse(is.na(.), NA, ./RowMean)))

write.csv(single_cell_again, "Output files/single_cell_again.csv")

single_cell_again <- single_cell_again %>% #add CT to all cell types to improve pivot
  rename_at(vars(-feature, -RowSum, -RowMean), 
            ~paste0("CT", .)) 

long_df <- single_cell_again %>%
  pivot_longer(cols = starts_with("CT"), 
               names_to = "cell_type", 
               values_to = "gene_expression") %>% 
  mutate(CellTypeAgg = str_extract(cell_type, ".{2}(.{4})")) %>%
  group_by(feature, CellTypeAgg) %>%
  mutate(CTMean = mean(gene_expression))

genes_not_in_long_df <- setdiff(outputforSTRING$genes, long_df$feature) #check if there are missing genes

#some genes have different names in this dataset to our original dataset. ensure all appropriate genes are included - our peptide = their gene. ATPF1D = ATP5D. H2AW = HIST3H2A. H3-3A = H3F3A. change these gene names in the original dataframe

long_df <- long_df %>%
  mutate(feature2 = feature) %>%
  mutate(feature2 = str_replace(feature2, "ATP5D", "ATP5F1D")) %>%
  mutate(feature2 = str_replace(feature2, "GIF", "MIF")) %>%
  mutate(feature2 = str_replace(feature2, "HIST3H2A", "H2AW")) #change names so they're recognised on merge. 1 gene (IGHV3-48) not in their dataset

long_df_cognition <- long_df %>%
  filter(feature2 %in% outputforSTRING$genes)

#get an aggregate estimate across peptides from the same gene
cogvspath_significant <- merge(cogvspath_significant, nice_table2[, c("FullPeptide", "GeneName")], by = "FullPeptide", all.x = TRUE)
cogvspath_significant <- cogvspath_significant %>%
  unique()

cognition_cont_significant <- merge(cognition_cont_significant, nice_table2[, c("FullPeptide", "GeneName")], by = "FullPeptide", all.x = TRUE)
cognition_cont_significant <- cognition_cont_significant %>%
  unique()

cognition_cont_significant_STRING <- cognition_cont_significant %>%
  filter(GeneName %in% outputforSTRING$genes)

allpeptides_cell_seq <- outputforSTRING %>%
  group_by(genes) %>%
  mutate(peptide_mean = mean(estimate.y)) %>% #takes mean of coefficients from continuous lm
  select(genes, peptide_mean) %>%
  unique()  

long_df_cognition_grouped <- long_df_cognition %>%
  filter(!CellTypeAgg %in% c('CTEndo', 'CTVLMC', 'CTPeri', 'CTOPC')) %>%
  filter(!is.na(CellTypeAgg)) %>%
  group_by(CellTypeAgg)

allpeptides_cell_seq2 <- merge(allpeptides_cell_seq, long_df_cognition_grouped, by.x = "genes", by.y = "feature2")

library(ggrepel)
library(ggplot2)

allpeptides_excitatory <- allpeptides_cell_seq2 %>%
  filter(CellTypeAgg== 'CTExc.') %>%
  select(genes, peptide_mean, CellTypeAgg, CTMean) %>%
  unique()

excplot <- ggplot(data = allpeptides_excitatory, aes(x = peptide_mean, y = CTMean, label = genes, color = peptide_mean > 0)) +
  geom_point() +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_text_repel(size = 5, vjust = 2, max.overlaps = 8) +
  scale_y_continuous(limits = c(0, 2.5)) +
  scale_color_manual(values = c("magenta3", "darkgreen")) +  # Assign colors based on condition
  theme_classic() + theme(legend.position = "none") +
  labs(x = "Mean peptide coefficient per gene", y = "Mean enrichment count", title = "Gene expression in excitatory neurons") +
  theme(text = element_text(size = 15), axis.title = element_text(size = 15),
        plot.title = element_text(size = 17))

print(excplot)

ggsave("Figures/excplot.pdf", plot = excplot, device = "pdf", width = 7, height = 8, units = "in", dpi = 300)

important_peps <- allpeptides_cell_seq2 %>%
  filter(feature %in% c("ITSN1", "CHGA", "CCK", "CHGB", "NEFM", "HNRNPA2B1", "SST", "TAC1", "VGF"))

allpeptides_inhibitory <- allpeptides_cell_seq2 %>%
  filter(CellTypeAgg== 'CTInh.') %>%
  select(genes, peptide_mean, CellTypeAgg, CTMean) %>%
  unique()

inhplot <- ggplot(data = allpeptides_inhibitory, aes(x = peptide_mean, y = CTMean, label = genes, color = peptide_mean > 0)) +
  geom_point() +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_text_repel(size = 5, vjust = 2, max.overlaps = 20) +
  scale_y_continuous(limits = c(0, 2.5)) +
  scale_color_manual(values = c("magenta3", "darkgreen")) + 
  theme_classic() + theme(legend.position = "none") +
  labs(x = "Mean peptide coefficient per gene", y = "Mean enrichment count", title = "Gene expression in inhibitory neurons") +
  theme(axis.title = element_text(size = 15),
        plot.title = element_text(size = 17))

print(inhplot)

ggsave("Figures/inhplot.pdf", plot = inhplot, device = "pdf", width = 7, height = 8, units = "in", dpi = 300)

```

**Specific cell-type analysis**
```{r}

#inhibitory. remove all cell types for which no expression of the proteins of interest 

long_df_cognition_inhib <- long_df_cognition %>%
    filter(CellTypeAgg== 'CTInh.') %>%
    filter(feature2 %in% c("SST", "VGF", "TAC1"))

layers <- ggplot(data = long_df_cognition_inhib, aes(x = cell_type, y = gene_expression)) +
  geom_bar(stat = "identity") +
  labs(x = "Cell Type", y = "Gene Expression") +
  ggtitle("Gene Expression by Cell Type in Layers") +
  facet_wrap(~feature2, scales = "free") +  theme_classic() +
  theme(plot.title = element_text(face = "bold", size = 15), strip.text = element_text(size = 25), axis.text.x = element_text(angle = 90, hjust = 1, size = 10), axis.text.y = element_text(size = 15))
ggsave("TAC1layers.pdf", plot = layers, device = "pdf", width = 20, height = 10, units = "in", dpi = 300)
ggsave("Figures/TAC1layers.pdf")

inhsignalling <- ggplot(long_df_cognition_inhib, aes(x = cell_type, y = gene_expression, fill = feature)) +
  geom_bar(stat = "identity", position = position_dodge2(width = 0.9, preserve = "single")) +
  geom_text(aes(label = feature), position = position_dodge(width = 0.9), vjust = -0.5, size = 3, color = "black") +
  labs(
    title = "Gene Expression of SST, VGF, and TAC1 in Different Inhibitory Cell Types",
    x = "Cell Type",
    y = "Expression Value"
  ) +
  scale_fill_manual(values = c("SST" = "red", "VGF" = "blue", "TAC1" = "green")) +
  theme_minimal() +
  coord_flip()
ggsave("inhsignalling.pdf", plot = inhsignalling, device = "pdf", width = 20, height = 20, units = "in", dpi = 300)

#inhib heatmap

long_df_cognition_inhib_cluster <- long_df_cognition_inhib %>%
    ungroup() %>%
    select(cell_type, gene_expression, feature) %>%
    pivot_wider(names_from = feature, values_from = gene_expression) %>%
    rowwise() %>% 
    filter(sum(c(SST,TAC1,VGF)) != 0) %>%
    column_to_rownames(var = "cell_type")

long_df_cognition_inhib_cluster <- as.matrix(long_df_cognition_inhib_cluster)

library(RColorBrewer)
pdf(file="Figures/long_df_cognition_inhib_cluster_plot2.pdf", width = 10, height = 14)
heatmap.2(long_df_cognition_inhib_cluster, col = brewer.pal(9, "Blues"), trace = "none", density.info = "none", dendrogram = "row", keysize = 0.8, cexRow = 1.3, margins = c(8, 16))
dev.off()


#excite heatmap

long_df_cognition_excite <- long_df_cognition %>%
    filter(CellTypeAgg== 'CTExc.') %>%
    filter(feature2 %in% c("CRYM", "DBN1", "NEFM", "PALM", "TUBA8"))

long_df_cognition_excite_cluster <- long_df_cognition_excite %>%
    ungroup() %>%
    select(cell_type, gene_expression, feature) %>%
    pivot_wider(names_from = feature, values_from = gene_expression) %>%
    rowwise() %>% 
    filter(sum(c(CRYM,DBN1,NEFM,PALM,TUBA8)) != 0) %>%
    column_to_rownames(var = "cell_type")

long_df_cognition_excite_cluster <- as.matrix(long_df_cognition_excite_cluster)


#get longer colour palette for plotting
par(mar=c(0,0,1,0))
coul <- brewer.pal(9, "Purples") 
coul <- colorRampPalette(coul)(40)
pie(rep(1, length(coul)), col = coul , main="") 


pdf(file="Figures/long_df_cognition_excite_cluster_plot2.pdf", width = 10, height = 14)
heatmap.2(long_df_cognition_excite_cluster, col = coul, trace = "none", density.info = "none", dendrogram = "row", keysize = 1, cexRow = 1.1, margins = c(8, 16))
dev.off()

```